File: src/data_structures/array/merge_sorted_test.php
<?php declare(strict_types=1);

namespace Ehunov\Interview\data_structures\array;

use PHPUnit\Framework\TestCase;

final class merge_sorted_test extends TestCase
{
    /**
     * @dataProvider provideTestMergeData
     */
    public function testMerge(array $nums1, int $n, array $nums2, int $m, array $expected): void
    {
        $this->assertEquals($expected, merge_sorted::merge($nums1, $n, $nums2, $m));
    }

    public function provideTestMergeData(): iterable
    {
        yield [
            [1, 2, 3, 0, 0, 0],
            3,
            [2, 5, 6],
            3,
            [1, 2, 2, 3, 5, 6],
        ];
        yield [
            [1],
            1,
            [],
            0,
            [1],
        ];
        yield [
            [0],
            0,
            [1],
            1,
            [1],
        ];
        yield [
            [2, 0],
            1,
            [1],
            1,
            [1, 2],
        ];
    }
}

File: src/data_structures/array/merge_sorted_test.go
package array

import (
	"reflect"
	"testing"
)

func TestMerge(t *testing.T) {
	tests := []struct {
		nums1    []int
		m        int
		nums2    []int
		n        int
		expected []int
	}{
		{
			[]int{1, 2, 3, 0, 0, 0},
			3,
			[]int{2, 5, 6},
			3,
			[]int{1, 2, 2, 3, 5, 6},
		}, {
			[]int{1},
			1,
			[]int{},
			0,
			[]int{1},
		}, {
			[]int{0},
			0,
			[]int{1},
			1,
			[]int{1},
		},
		{
			[]int{2, 0},
			1,
			[]int{1},
			1,
			[]int{1, 2},
		},
	}
	for i, tt := range tests {
		t.Run(string(rune(i)), func(t *testing.T) {
			if got := merge(tt.nums1, tt.m, tt.nums2, tt.n); !reflect.DeepEqual(got, tt.expected) {
				t.Errorf("merge() = %v, expected %v", got, tt.expected)
			}
		})
	}
}
